target/linux/ramips/image/common-sercomm.mk

```diff
@@ -16,6 +17,24 @@ define Build/sercomm-crypto
 	rm -f $@.enc $@.key
 endef
 
+define Build/sercomm-factory-cqr
+	$(call Build/sercomm-kernel-trim,$(IMAGE_KERNEL) \
+		$(IMAGE_KERNEL).data)
+	$(call Build/sercomm-pid,$@.hdrfactory)
+	$(call Build/sercomm-write-pid,$(IMAGE_KERNEL).data $$((0x70)) \
+		$@.hdrfactory)
+	$(call Build/sercomm-write-pid,$@ $$((0x80)) $@.hdrfactory)
+	cat $(IMAGE_KERNEL).data $@ | $(MKHASH) md5 | \
+		awk '{print $$1}' | tr -d '\n' | \
+		dd seek=$$((0x1e0)) of=$@.hdrfactory bs=1 \
+		conv=notrunc 2>/dev/null
+	$(call Build/sercomm-kernel-factory,$(IMAGE_KERNEL).data \
+		$@ $@.hdrkrnl)
+	cat $@.hdrfactory $@.hdrkrnl $(IMAGE_KERNEL).data $@ > $@.new
+	mv $@.new $@
+	rm -f $@.hdrfactory $@.hdrkrnl $(IMAGE_KERNEL).data
```

Comment:
I do not like all these `call Build` commands. 

I think you should add the commands modifying the kernel image here:
````
  KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | lzma -a0 | \
	uImage lzma | sercomm-kernel
````

Maybe line this:
````
  KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | lzma -a0 | \
	uImage lzma | sercomm-kernel | sercomm-kernel-trim | sercomm-write-pid | sercomm-kernel-factory
````
(This is not directly working like this)

Reply:
@hauke thanks for taking the time to look at these scripts. The essence of the problem that we need 2 different structures for the factory and sysupgrade images:
sysupgrade: `kernel header 1 | kernel `
factory:  `kernel header 1 | kernel header 2 | kernel `

Headers data depends on kernel data, ubi rootfs data and their offsets (which are different for the each Slot of the firmware)

This is correct for the sysupgrade (so, we can't touch this):
```
KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | lzma -a0 | \
	uImage lzma | sercomm-kernel
```

We have to insert `kernel header 2` between `kernel header 1` and `kernel ` for the factory with correct checksums and so on. Also, factory image has a first header containing checksum of combined pure kernel + rootfs. That's why scripts are so complicated.

Perhaps I did not understood something. Could you explain your idea regarding kernel recipe in more detail, please?

Reply:
Let's up this 2 yo PR. 🥇 

@hauke do we have any merge blockers except these `call Build` commands?

Reply:
Hi!

> This is correct for the sysupgrade (so, we can't touch this):
> 
> ```
> KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | lzma -a0 | \
> 	uImage lzma | sercomm-kernel
> ```

You could drop the sercomm-kernel from KERNEL and apply it in factory.bin and sysupgrade.bin recipe separately. This way you can describe the recipe with steps, and 

> We have to insert `kernel header 2` between `kernel header 1` and `kernel ` for the factory with correct checksums and so on. Also, factory image has a first header containing checksum of combined pure kernel + rootfs. That's why scripts are so complicated.

This also eliminates the need to erase the existing `kernel header 1` before creating the factory image, as you are just adding the two header in factory.bin recipe.

Reply:
Hi @981213! Thanks for your idea. What do you think about this?
```
diff --git a/target/linux/ramips/image/common-sercomm.mk b/target/linux/ramips/image/common-sercomm.mk
index 83bb2a742c..54f232fc6b 100644
--- a/target/linux/ramips/image/common-sercomm.mk
+++ b/target/linux/ramips/image/common-sercomm.mk
@@ -16,6 +16,23 @@ define Build/sercomm-crypto
 	rm -f $@.enc $@.key
 endef
 
+define Build/sercomm-factory-cqr
+	$(TOPDIR)/scripts/sercomm-pid.py \
+		--hw-version $(SERCOMM_HWVER) \
+		--hw-id $(SERCOMM_HWID) \
+		--sw-version $(SERCOMM_SWVER) \
+		--pid-file $@.fhdr
+	printf $$(stat -c%s $(IMAGE_KERNEL)) | \
+		dd seek=$$((0x70)) of=$@.fhdr bs=1 conv=notrunc 2>/dev/null
+	printf $$(($$(stat -c%s $@)-$$(stat -c%s $(IMAGE_KERNEL))-$$((0x200)))) | \
+		dd seek=$$((0x80)) of=$@.fhdr bs=1 conv=notrunc 2>/dev/null
+	dd bs=$$((0x200)) skip=1 if=$@ conv=notrunc 2>/dev/null | \
+		$(MKHASH) md5 | awk '{print $$1}' | tr -d '\n' | \
+		dd seek=$$((0x1e0)) of=$@.fhdr bs=1 conv=notrunc 2>/dev/null
+	dd if=$@ >> $@.fhdr 2>/dev/null
+	mv $@.fhdr $@
+endef
+
 define Build/sercomm-kernel
 	$(TOPDIR)/scripts/sercomm-kernel-header.py \
 		--kernel-image $@ \
@@ -26,6 +43,22 @@ define Build/sercomm-kernel
 	mv $@.hdr $@
 endef
 
+define Build/sercomm-kernel-factory
+	$(TOPDIR)/scripts/sercomm-kernel-header.py \
+		--kernel-image $@ \
+		--kernel-offset $(SERCOMM_KERNEL_OFFSET) \
+		--rootfs-offset $(SERCOMM_ROOTFS_OFFSET) \
+		--output-header $@.khdr1
+	$(TOPDIR)/scripts/sercomm-kernel-header.py \
+		--kernel-image $@ \
+		--kernel-offset $(SERCOMM_KERNEL2_OFFSET) \
+		--rootfs-offset $(SERCOMM_ROOTFS2_OFFSET) \
+		--output-header $@.khdr2
+	cat $@.khdr1 $@.khdr2 > $@.khdr
+	dd if=$@ >> $@.khdr 2>/dev/null
+	mv $@.khdr $@
+endef
+
 define Build/sercomm-part-tag
 	$(call Build/sercomm-part-tag-common,$(word 1,$(1)) $@)
 endef
@@ -64,6 +97,28 @@ define Build/sercomm-prepend-tagged-kernel
 	mv $(IMAGE_KERNEL).tagged $@
 endef
 
+define Device/sercomm_cxx
+  $(Device/dsa-migration)
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  UBINIZE_OPTS := -E 5
+  LOADER_TYPE := bin
+  KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | lzma -a0 | \
+	uImage lzma
+  IMAGES += factory.img
+  IMAGE/sysupgrade.bin := append-kernel | sercomm-kernel | \
+	sysupgrade-tar kernel=$$$$@ | append-metadata
+  KERNEL_SIZE := 6144k
+  KERNEL_LOADADDR := 0x81001000
+  LZMA_TEXT_START := 0x82800000
+  SERCOMM_KERNEL_OFFSET := 0x400100
+  SERCOMM_ROOTFS_OFFSET := 0x1000000
+  SERCOMM_KERNEL2_OFFSET := 0xa00100
+  SERCOMM_ROOTFS2_OFFSET := 0x3000000
+  IMAGE/factory.img := append-kernel | sercomm-kernel-factory | \
+	append-ubi | sercomm-factory-cqr
+endef
+
 define Device/sercomm_dxx
   $(Device/dsa-migration)
   BLOCKSIZE := 128k

```

Reply:
> What do you think about this?

Yeah, this one is better :D